From: Russ Allbery <rra@debian.org>
Date: Thu, 17 Jan 2019 19:21:40 -0800
Subject: Verify scp command options

ESnet discovered a security vulnerability in the scp backend for
rssh.  Since the arguments to scp on the server side were not
checked, the client could pass in arbitrary scp command-line flags,
including setting arbitrary scp options.  This allows setting the
option PKCS11Provider, which loads and executes code from a shared
module.

Even if the -o flag is blocked, this is still possible via -F to
load an already-uploaded ssh configuration file, or, if .ssh/config
is writable, by just uploading that configuration file directly
first.

Attempt to protect against this attack by checking the command line
of scp and only allowing the options that are passed to the server
end of the connection.  Specifically, do not allow multiple
non-option arguments, which attempts to prevent causing the server
to initiate an scp command.  (This will break scp -3 through rssh,
which seems like an acceptable tradeoff.)

Debian Bug#919623
---
 util.c | 46 ++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 44 insertions(+), 2 deletions(-)

diff --git a/util.c b/util.c
index 3ec01e1..6e668c9 100644
--- a/util.c
+++ b/util.c
@@ -264,6 +264,45 @@ static int rsync_okay( char **vec )
 }
 
 
+/*
+ * scp_okay() - take the command line and check that it is a hopefully-safe scp
+ *		server command line, accepting only very specific options.
+ *		Returns FALSE if the command line should not be allowed, TRUE
+ *		if it is okay.
+ */
+static int scp_okay( char **vec )
+{
+	int saw_file = FALSE;
+	int saw_end  = FALSE;
+
+	for ( vec++; vec && *vec; vec++ ){
+		/* Allowed options. */
+		if ( !saw_end ) {
+			if ( strcmp(*vec, "-v") == 0 ) continue;
+			if ( strcmp(*vec, "-r") == 0 ) continue;
+			if ( strcmp(*vec, "-p") == 0 ) continue;
+			if ( strcmp(*vec, "-d") == 0 ) continue;
+			if ( strcmp(*vec, "-f") == 0 ) continue;
+			if ( strcmp(*vec, "-t") == 0 ) continue;
+		}
+
+		/* End of arguments.  One more argument allowed after this. */
+		if ( !saw_end && strcmp(*vec, "--") == 0 ){
+			saw_end = TRUE;
+			continue;
+		}
+
+		/* No other options allowed, but allow file starting with -. */
+		if ( *vec[0] == '-' && !saw_end ) return FALSE;
+		if ( saw_file ) return FALSE;
+		saw_file = TRUE;
+	}
+
+	/* We must have seen a single file. */
+	return saw_file;
+}
+
+
 /*
  * check_command_line() - take the command line passed to rssh, and verify
  *			  that the specified command is one the user is
@@ -279,8 +318,11 @@ char *check_command_line( char **cl, ShellOptions_t *opts )
 		return PATH_SFTP_SERVER;
 
 	if ( check_command(*cl, opts, PATH_SCP, RSSH_ALLOW_SCP) ){
-		/* filter -S option */
-		if ( opt_filter(cl, 'S') ) return NULL;
+		if ( !scp_okay(cl) ){
+			fprintf(stderr, "\ninsecure scp option not allowed.");
+			log_msg("insecure scp option in scp command line");
+			return NULL;
+		}
 		return PATH_SCP;
 	}
 
